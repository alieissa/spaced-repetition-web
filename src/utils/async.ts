/** @format */
/**
 * Credit goes to https://medium.com/@fillopeter/pattern-matching-with-typescript-done-right-94049ddd671c
 * For the pattern matching code
 */
export type Untriggered = {
  type: 'Untriggered'
}
export function Untriggered(): Untriggered {
  return {
    type: 'Untriggered',
  }
}

const t: Optional<Answer, 'easiness' | 'quality' | 'interval'> = {
  id: '',
  content: 'fsdfds',
}
export type Loading<V> = {
  type: 'Loading'
  value: V
}
export function Loading<V>(v: V): Loading<V> {
  return {
    type: 'Loading',
    value: v,
  }
}

export type Failure<E> = {
  type: 'Failure'
  value: E
}
export function Failure<E>(e: E): Failure<E> {
  return {
    type: 'Failure',
    value: e,
  }
}

export type Success<V> = {
  type: 'Success'
  value: V
}
export function Success<V>(v: V): Success<V> {
  return {
    type: 'Success',
    value: v,
  }
}

export type Async<L = unknown, F = unknown, S = unknown> =
  | Untriggered
  | Loading<L>
  | Failure<F>
  | Success<S>

export type AsyncStatus<R, E, A> = Async<R, E, A>['type']

export type AsyncMap<R, E, A, U> = {
  [K in AsyncStatus<R, E, A>]: U extends { type: K } ? U : never
}

export type AsyncStatusMap<R, E, A> = AsyncMap<R, E, A, Async<R, E, A>>
export type AsyncPattern<T, R, E, A> = {
  [K in keyof AsyncStatusMap<R, E, A>]: (a: AsyncStatusMap<R, E, A>[K]) => T
}

export function match<R, E, A>(
  e: Async<R, E, A>,
): <T>(pattern: AsyncPattern<T, R, E, A>) => T {
  return <T>(pattern: AsyncPattern<T, R, E, A>) => pattern[e.type](e as any)
}
