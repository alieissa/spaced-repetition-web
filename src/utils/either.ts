/** @format */

/**
 * Credit goes to https://medium.com/@fillopeter/pattern-matching-with-typescript-done-right-94049ddd671c
 * For the pattern matching code
 */
export type Right<V> = Readonly<{
  type: 'Right'
  value: V
}>
export function Right<V>(v: V): Right<V> {
  return {
    type: 'Right',
    value: v,
  }
}

export type Left<E = Error> = Readonly<{
  type: 'Left'
  value: E
}>
export function Left<E>(e: E): Left<E> {
  return {
    type: 'Left',
    value: e,
  }
}

export type Either<L, R> = Left<L> | Right<R>

export type EitherType<L, R> = Either<L, R>['type']
export type EitherMap<L, R, U> = {
  [K in EitherType<L, R>]: U extends { type: K } ? U : never
}
export type EitherTypeMap<L, R> = EitherMap<L, R, Either<L, R>>
export type EitherPattern<T, L, R> = {
  [K in keyof EitherTypeMap<L, R>]: (either: EitherTypeMap<L, R>[K]) => T
}

export function match<L, R>(
  e: Either<L, R>,
): <T>(pattern: EitherPattern<T, L, R>) => T {
  return <T>(pattern: EitherPattern<T, L, R>) => pattern[e.type](e as any)
}
