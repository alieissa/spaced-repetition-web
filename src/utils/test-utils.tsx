/** @format */

import { combineReducers, configureStore } from '@reduxjs/toolkit'
import { render } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import { RouterProvider, createMemoryRouter } from 'react-router-dom'
import decksReducer from 'src/modules/decks/decks.reducer'

// As a basic setup, import your same slice reducers
import { loginReducer } from 'src/modules/login'
import { signupReducer } from 'src/modules/signup'
import { verificationReducer } from 'src/modules/verification'

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
// interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
//   preloadedState?: PreloadedState<RootState>
//   store?: AppStore
// }

const reducers = combineReducers({
  login: loginReducer,
  signup: signupReducer,
  verification: verificationReducer,
  decks: decksReducer,
})

const withRedux = (ui: React.ReactElement) => {
  const store = configureStore({ reducer: reducers })
  return <Provider store={store}>{ui}</Provider>
}
const withRouter = (ui: React.ReactElement) => {
  const router = createMemoryRouter([{ element: ui, path: '/' }], {
    initialEntries: ['/'],
    initialIndex: 1,
  })

  return <RouterProvider router={router} />
}
export function renderWithProviders(ui: React.ReactElement) {
  // Return an object with the store and all of RTL's query functions
  return render(withRedux(withRouter(ui)))
}

export const flushPromises = () =>
  new Promise((resolve) => setTimeout(resolve, 0))
