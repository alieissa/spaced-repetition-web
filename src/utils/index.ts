/** @format */

import * as async from './async'
import * as either from './either'

export { async }
export { either }
