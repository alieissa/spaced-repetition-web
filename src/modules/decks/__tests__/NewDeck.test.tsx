/** @format */

import '@testing-library/jest-dom'

import { renderWithProviders } from 'src/utils/test-utils'
import { NewDeck } from '..'

describe('NewDeck', () => {
  describe('view', () => {
    it('should render correctly', () => {
      const { asFragment } = renderWithProviders(<NewDeck />)
      expect(asFragment()).toMatchSnapshot()
    })
  })
})
