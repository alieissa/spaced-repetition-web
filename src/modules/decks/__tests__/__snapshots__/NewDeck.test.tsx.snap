// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewDeck view should render correctly 1`] = `
<DocumentFragment>
  <div
    class="ui container"
    data-testid="deck-success"
  >
    <div
      class="ui basic segment"
      style="padding: 0px;"
    >
      <header
        class="justify-space-between"
      >
        <h2>
          Create deck
        </h2>
        <div>
          <button
            class="ui small button mx-auto"
          >
            Cancel
          </button>
          <button
            class="ui green small button"
            data-testid="deck-save"
          >
            Save
          </button>
        </div>
      </header>
    </div>
    <main>
      <div>
        <div
          class="ui fluid card"
        >
          <div
            class="content bt-none"
            style="padding: 0px;"
          >
            <div
              class="ui list"
              role="list"
            >
              <div
                class="item"
                role="listitem"
              >
                <div
                  class="ui input w-full"
                  data-testid="deck-name"
                >
                  <input
                    placeholder="Enter name here"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class="item"
                role="listitem"
              >
                <div
                  class="ui input w-full"
                  data-testid="deck-description"
                >
                  <input
                    placeholder="Enter description here"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <section
          class="flex-column w-inherit test"
        >
          <div
            class="ui list"
            role="list"
          >
            <div
              class="item"
              role="listitem"
            >
              <div
                class="ui segment bs-0"
              >
                <form
                  class="ui form w-full"
                >
                  <div
                    class="ui horizontal list flex"
                    role="list"
                    style="display: flex;"
                  >
                    <div
                      class="item flex-1"
                      role="listitem"
                    >
                      <div
                        class="ui input w-full"
                      >
                        <input
                          name="question-content"
                          placeholder="Enter question here"
                          type="text"
                          value=""
                        />
                      </div>
                    </div>
                    <div
                      class="item flex-1"
                      role="listitem"
                    >
                      <div
                        class="ui list"
                        role="list"
                        style="padding: 0px;"
                      >
                        <div
                          class="item flex"
                          role="listitem"
                          style="display: flex;"
                        >
                          <div
                            class="ui input w-full"
                          >
                            <input
                              name="answer-content"
                              placeholder="Enter answer here"
                              type="text"
                              value=""
                            />
                          </div>
                          <button
                            class="ui small icon disabled button"
                            disabled=""
                            style="background-color: white;"
                            tabindex="-1"
                          >
                            <i
                              aria-hidden="true"
                              class="x icon"
                            />
                          </button>
                        </div>
                        <div
                          class="item"
                          role="listitem"
                          style="text-align: right;"
                        >
                          <button
                            class="ui small icon button"
                            style="background-color: white;"
                          >
                            <i
                              aria-hidden="true"
                              class="green plus icon"
                            />
                          </button>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="fields"
                    style="justify-content: flex-end;"
                  />
                </form>
              </div>
            </div>
          </div>
        </section>
        <div
          class="ui basic segment flex-row-reverse test"
          style="padding: 0px;"
        >
          <button
            class="ui green icon button"
          >
            <i
              aria-hidden="true"
              class="plus icon"
            />
          </button>
        </div>
      </div>
    </main>
  </div>
</DocumentFragment>
`;
